diff -ru wasm-micro-runtime-WAMR-2.4.2.orig/build-scripts/build_llvm.py wasm-micro-runtime-WAMR-2.4.2/build-scripts/build_llvm.py
--- wasm-micro-runtime-WAMR-2.4.2.orig/build-scripts/build_llvm.py	2025-10-11 12:28:16.653222239 +0800
+++ wasm-micro-runtime-WAMR-2.4.2/build-scripts/build_llvm.py	2025-10-11 12:28:48.276741632 +0800
@@ -236,6 +236,11 @@
 def main():
     parser = argparse.ArgumentParser(description="build necessary LLVM libraries")
     parser.add_argument(
+        "--deps-dir",
+        type=str,
+        default="",
+    )
+    parser.add_argument(
         "--platform",
         type=str,
         choices=["android", "arc", "darwin", "linux", "windows", "xtensa"],
@@ -319,8 +324,11 @@
     if current_file.is_symlink():
         current_file = pathlib.Path(os.readlink(current_file))
 
-    current_dir = current_file.parent.resolve()
-    deps_dir = current_dir.joinpath("../core/deps").resolve()
+    if options.deps_dir:
+        deps_dir = pathlib.Path(options.deps_dir)
+    else:
+        current_dir = current_file.parent.resolve()
+        deps_dir = current_dir.joinpath("../core/deps").resolve()
 
     try:
         llvm_info = llvm_repo_and_branch.get(platform, llvm_repo_and_branch["default"])
diff -ru wasm-micro-runtime-WAMR-2.4.2.orig/build-scripts/config_common.cmake wasm-micro-runtime-WAMR-2.4.2/build-scripts/config_common.cmake
--- wasm-micro-runtime-WAMR-2.4.2.orig/build-scripts/config_common.cmake	2025-10-11 12:28:16.656555593 +0800
+++ wasm-micro-runtime-WAMR-2.4.2/build-scripts/config_common.cmake	2025-10-11 12:29:31.690330374 +0800
@@ -100,19 +100,19 @@
     set (WAMR_BUILD_LAZY_JIT 1)
   endif ()
   if (NOT DEFINED LLVM_DIR)
-    set (LLVM_SRC_ROOT "${WAMR_ROOT_DIR}/core/deps/llvm")
+    set (LLVM_SRC_ROOT "${DEPS_DIR}/llvm")
     set (LLVM_BUILD_ROOT "${LLVM_SRC_ROOT}/build")
     if (NOT EXISTS "${LLVM_BUILD_ROOT}")
       message(STATUS "LLVM build directory not found, attempting to build LLVM...")
       if (DEFINED LLVM_CMAKE_FLAGS)
         execute_process(
-          COMMAND python3 ${WAMR_ROOT_DIR}/build-scripts/build_llvm.py --extra-cmake-flags ${LLVM_CMAKE_FLAGS}
+          COMMAND python3 ${WAMR_ROOT_DIR}/build-scripts/build_llvm.py --deps-dir ${DEPS_DIR} --extra-cmake-flags ${LLVM_CMAKE_FLAGS}
           WORKING_DIRECTORY ${WAMR_ROOT_DIR}
           RESULT_VARIABLE result
         )
       else()
         execute_process(
-          COMMAND python3 ${WAMR_ROOT_DIR}/build-scripts/build_llvm.py
+          COMMAND python3 ${WAMR_ROOT_DIR}/build-scripts/build_llvm.py --deps-dir ${DEPS_DIR}
           WORKING_DIRECTORY ${WAMR_ROOT_DIR}
           RESULT_VARIABLE result
         )
